client/client.go

```diff
@@ -1,44 +1,123 @@
+func NewClient(endpoint string, logger *logger.Logger) *Client {
+	return &Client{
+		httpClient: http.DefaultClient,
+		endpoint:   endpoint,
+		logger:     logger,
 	}
-
-	return client
 }
 
 // NewAuthorizedRequest returns a new GraphQL request with the
 // authorization headers set for CircleCI auth.
-func NewAuthorizedRequest(token, query string) *graphql.Request {
-	req := graphql.NewRequest(query)
+func NewAuthorizedRequest(token, query string) *Request {
+	req := NewRequest(query)
 	req.Header.Set("Authorization", token)
 	req.Header.Set("User-Agent", version.UserAgent())
 	return req
 }
 
-// Run will construct a request using graphql.NewRequest.
-// Then it will execute the given query using graphql.Client.Run.
-// This function will return the unmarshalled response as JSON.
-func Run(client *graphql.Client, token, query string) (map[string]interface{}, error) {
-	req := NewAuthorizedRequest(token, query)
-	ctx := context.Background()
-	var resp map[string]interface{}
-	err := client.Run(ctx, req, &resp)
-	return resp, err
+// Request is a GraphQL request.
+type Request struct {
+	Query     string                 `json:"query"`
+	Variables map[string]interface{} `json:"variables"`
+
+	// Header represent any request headers that will be set
+	// when the request is made.
+	Header http.Header `json:"-"`
+}
+
+// NewRequest makes a new Request with the specified query string.
+func NewRequest(query string) *Request {
+	request := &Request{
+		Query:     query,
+		Variables: make(map[string]interface{}),
+		Header:    make(map[string][]string),
+	}
+	return request
+}
+
+// Var sets a variable.
+func (request *Request) Var(key string, value interface{}) {
+	request.Variables[key] = value
+}
+
+func (c *Client) prepareRequest(ctx context.Context, request *Request) (*http.Request, error) {
+	var requestBody bytes.Buffer
+	if err := json.NewEncoder(&requestBody).Encode(request); err != nil {
+		return nil, errors.Wrap(err, "encode body")
+	}
+	c.logger.Debug(">> variables: %v", request.Variables)
+	c.logger.Debug(">> query: %s", request.Query)
+	r, err := http.NewRequest(http.MethodPost, c.endpoint, &requestBody)
+	if err != nil {
+		return nil, err
+	}
+	r.Header.Set("Content-Type", "application/json; charset=utf-8")
+	r.Header.Set("Accept", "application/json; charset=utf-8")
+	for key, values := range request.Header {
+		for _, value := range values {
+			r.Header.Add(key, value)
+		}
+	}
+
+	r = r.WithContext(ctx)
+	return r, nil
+}
+
+// Run sends an HTTP request to the GraphQL server and deserializes the response or returns an error.
+func (c *Client) Run(ctx context.Context, request *Request, resp interface{}) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	req, err := c.prepareRequest(ctx, request)
+	if err != nil {
+		return err
+	}
+
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := res.Body.Close()
+		if err != nil {
+			log.Fatal(err)
```

Comment:
This will `os.Exit()` if there is an error closing the body. That's probably not good. Closing the body is not really critical, especially in a CLI where the process is short lived. I would suggest either printing a warning  or ignoring it all together.